backend/api/src/subscription.rs

```diff
@@ -0,0 +1,161 @@
+use std::{
+    collections::HashMap,
+    ops::{Deref, DerefMut},
+};
+
+use anyhow::Result;
+use common::context::Context;
+use fred::{clients::SubscriberClient, prelude::PubsubInterface, types::RedisValue};
+use tokio::{
+    select,
+    sync::{broadcast, mpsc, oneshot, Mutex},
+};
+
+#[derive(Debug)]
+enum Event {
+    Subscribe {
+        topic: String,
+        tx: oneshot::Sender<broadcast::Receiver<RedisValue>>,
+    },
+    Unsubscribe {
+        topic: String,
+    },
+}
+
+pub struct SubscriptionManager {
+    events_tx: mpsc::UnboundedSender<Event>,
+    events_rx: Mutex<mpsc::UnboundedReceiver<Event>>,
+}
+
+impl Default for SubscriptionManager {
+    fn default() -> Self {
+        // Only one value is needed in the channel.
+        // This is a way to get around we cannot await in a drop.
+        let (events_tx, events_rx) = mpsc::unbounded_channel();
+
+        Self {
+            events_rx: Mutex::new(events_rx),
+            events_tx,
+        }
+    }
+}
+
+pub struct SubscriberReceiver<'a> {
+    topic: String,
+    rx: broadcast::Receiver<RedisValue>,
+    manager: &'a SubscriptionManager,
+}
+
+impl Deref for SubscriberReceiver<'_> {
+    type Target = broadcast::Receiver<RedisValue>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.rx
+    }
+}
+
+impl DerefMut for SubscriberReceiver<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.rx
+    }
+}
+
+impl SubscriptionManager {
+    pub async fn run(&self, ctx: Context, redis: SubscriberClient) -> Result<()> {
+        let mut handle = redis.manage_subscriptions();
+
+        let mut topics = HashMap::<String, broadcast::Sender<RedisValue>>::new();
+
+        let mut events_rx = self.events_rx.lock().await;
+
+        let mut messages = redis.on_message();
+
+        loop {
+            select! {
+                event = events_rx.recv() => {
+                    match event.unwrap() {
+                        Event::Subscribe { topic, tx } => {
+                            let topic = topic.to_lowercase();
```

Comment:
is there no better way to do this? why mutate?

Reply:
What do u mean?

Reply:
![image](https://github.com/ScuffleTV/scuffle/assets/18615981/b45076e0-a659-4fcf-98da-5715ca1de03e)
(from the discord conversation from context). essentially, i think it would be nice if we had a way to parse a string into a specific channel class which would absolutely concretely ensure the channel was correct so it would not have to be rechecked or reparsed into any other class again

Reply:
![image](https://github.com/ScuffleTV/scuffle/assets/18615981/a154d8a0-0594-478f-9d2d-fd60fadec57d)
alright this makes sense, resolving this now